File: scala-source/com/barrybecker4/simulation/liquid/LiquidDynamicOptions.scala
// Copyright by Barry G. Becker, 2016-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.liquid

import com.barrybecker4.ui.sliders.SliderGroup
import com.barrybecker4.ui.sliders.SliderGroupChangeListener
import com.barrybecker4.ui.sliders.SliderProperties

import javax.swing.*
import java.awt.*
import com.barrybecker4.simulation.liquid.model.UiParameter

import scala.collection.immutable.List


/**
  * Dynamic controls for the liquid simulation that will show on the right.
  * They change the behavior of the simulation while it is running.
  */
class LiquidDynamicOptions private[liquid](var liquidSim: LiquidSimulator)
  extends JPanel with SliderGroupChangeListener {

  setLayout(new BoxLayout(this, BoxLayout.Y_AXIS))
  setBorder(BorderFactory.createEtchedBorder)
  setPreferredSize(new Dimension(300, 300))

  private val uiParameters = liquidSim.getEnvironment.getUiParameters()
  private val sliderProperties = createSliderProperties(uiParameters)
  private val sliderGroup = new SliderGroup(sliderProperties)
  sliderGroup.setSliderListener(this)
  add(sliderGroup)
  
  val fill = new JPanel
  fill.setPreferredSize(new Dimension(10, 1000))
  add(fill)


  protected def createCheckBox(label: String, ttip: String, initialValue: Boolean): JCheckBox = {
    val cb = new JCheckBox(label, initialValue)
    cb.setToolTipText(ttip)
    cb
  }

  private def createSliderProperties(uiParameters: List[UiParameter]): Array[SliderProperties] = {
    uiParameters.map(parameter => {
      new SliderProperties(
        parameter.displayName,
        parameter.minValue,
        parameter.maxValue,
        parameter.initialValue,
        parameter.scale
      )
    }).toArray
  }

  def reset(): Unit = { sliderGroup.reset() }

  /** One of the sliders was moved. */
  override def sliderChanged(sliderIndex: Int, sliderName: String, value: Double): Unit = {
    if (sliderName == "time step (dt)") {
      liquidSim.setTimeStep(value)
    } else {
      val env = liquidSim.getEnvironment
      val fieldName = uiParameters(sliderIndex).name

      var currentClass: Class[_] = env.getClass
      var fieldFound = false

      while (currentClass != null && !fieldFound) { // search inheritance hierarchy
        try {
          val field = currentClass.getDeclaredField(fieldName)
          field.setAccessible(true)
          //println(s"Setting field $fieldName in ${currentClass.getSimpleName} = $value")
          field.set(env, value)
          fieldFound = true
        } catch {
          case _: NoSuchFieldException =>
            currentClass = currentClass.getSuperclass
        }
      }

      if (!fieldFound) {
        println(s"Could not find field '$fieldName' in class hierarchy of ${env.getClass.getSimpleName}")
      }
    }
  }
  
}

File: source/com/barrybecker4/simulation/liquid1/LiquidDynamicOptions.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.simulation.liquid1;

import com.barrybecker4.simulation.liquid1.compute.GridUpdater;
import com.barrybecker4.ui.sliders.SliderGroup;
import com.barrybecker4.ui.sliders.SliderGroupChangeListener;
import com.barrybecker4.ui.sliders.SliderProperties;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

/**
 * Dynamic controls for the liquid simulation that will show on the right.
 * They change the behavior of the simulation while it is running.
 * @author Barry Becker
 */
class LiquidDynamicOptions extends JPanel
                          implements SliderGroupChangeListener, ActionListener {

    private LiquidSimulator liquidSim;

    private static final String VISCOSITY_SLIDER = "Viscosity";
    private static final String B0_SLIDER = "Relaxation Coefficient";
    private static final String TIMESTEP_SLIDER = "Time Step Size";

    private SliderGroup sliderGroup;

    private JCheckBox advectionOnlyCheckBox;

    /**
     * Constructor
     */
    LiquidDynamicOptions(LiquidSimulator liquid) {

        setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
        setBorder(BorderFactory.createEtchedBorder());
        setPreferredSize(new Dimension(300, 300));

        liquidSim = liquid;

        sliderGroup = new SliderGroup(createSliderProperties());
        sliderGroup.addSliderChangeListener(this);

        add(sliderGroup);

        advectionOnlyCheckBox =
                createCheckBox("Do advection only",
                    "If checked we will not apply the Navier Stokes solver",
                    liquidSim.getAdvectionOnly() );
        add(advectionOnlyCheckBox);

        JPanel fill = new JPanel();
        fill.setPreferredSize(new Dimension(10, 1000));
        add(fill);
    }


    protected JCheckBox createCheckBox(String label, String ttip, boolean initialValue) {
        JCheckBox cb = new JCheckBox(label, initialValue);
        cb.setToolTipText(ttip);
        cb.addActionListener(this);
        return cb;
    }


    private SliderProperties[] createSliderProperties() {

        SliderProperties[] sliderProps;
        sliderProps = new SliderProperties[] {
                //                                      MIN   MAX    INITIAL    SCALE
                new SliderProperties(VISCOSITY_SLIDER, 0.0, 0.1, GridUpdater.DEFAULT_VISCOSITY, 100),
                new SliderProperties(B0_SLIDER, 1.0, 2.0, GridUpdater.DEFAULT_B0, 100),
                new SliderProperties(TIMESTEP_SLIDER, 0.001, 0.4, 0.01, 1000)};
         return sliderProps;
    }


    public void reset() {
        sliderGroup.reset();
    }

    /**
     * One of the sliders was moved.
     */
    public void sliderChanged(int sliderIndex, String sliderName, double value) {

        if (sliderName.equals(VISCOSITY_SLIDER)) {
            liquidSim.getEnvironment().setViscosity(value);
        }
        else if (sliderName.equals(B0_SLIDER)) {
            liquidSim.getEnvironment().setB0(value);
        }
        else if (sliderName.equals(TIMESTEP_SLIDER)) {
            liquidSim.setTimeStep(value);
        }
    }

    public void actionPerformed(ActionEvent e) {
        if (e.getSource() == advectionOnlyCheckBox) {
            liquidSim.setAdvectionOnly(advectionOnlyCheckBox.isSelected());
        }
    }
}
